lib/storage/src/Upload.ts

```diff
@@ -0,0 +1,185 @@
+  partNumber: number;
+  data: BodyDataTypes;
+}
+
+const MIN_PART_SIZE = 1024 * 1024 * 5;
+
+export class Upload extends EventEmitter {
+  /**
+   * S3 multipart upload does not allow more than 1000 parts.
+   */
+  private MAX_PARTS = 1000;
+
+  // Defaults
+  private queueSize = 4;
+  private partSize = MIN_PART_SIZE;
+  private leavePartsOnError = false;
+  private tags: Tag[] = [];
+
+  private client: ServiceClients;
+  private params: PutObjectCommandInput;
+
+  // used for reporting progress
+  private totalBytes?: number;
+  private bytesUploadedSoFar: number;
+
+  // used in the upload
+  private abortSignal: AbortController;
+  private concurrentUploaders: Promise<void>[] = [];
+
+  private uploadedParts: CompletedPart[] = [];
+  private uploadId?: string;
+
+  uploadEvent?: string;
+
+  constructor(options: Options) {
+    super();
+
+    // set defaults from options
+    this.queueSize = options.queueSize || this.queueSize;
+    this.partSize = options.partSize || this.partSize;
+    this.leavePartsOnError = options.leavePartsOnError || this.leavePartsOnError;
+    this.tags = options.tags || this.tags;
+
+    if (this.partSize < MIN_PART_SIZE) {
+      throw new Error(
+        `EntityTooSmall: Your proposed upload partsize [${this.partSize}] is smaller than the minimum allowed size [${MIN_PART_SIZE}] (5MB)`
+      );
+    }
+
+    if (this.queueSize < 1) {
+      throw new Error(`Queue size: Must have atleast one uploading queue.`);
+    }
+
+    this.client = options.client;
+    this.params = options.params;
+
+    // set progress defaults
+    this.totalBytes = byteLength(this.params.Body);
+    this.bytesUploadedSoFar = 0;
+    this.abortSignal = new AbortController();
+  }
+
+  async abort(): Promise<void> {
+    /**
+     * Abort stops all new uploads and immediately exists the top level promise on this.done()
+     * Concurrent threads in flight clean up eventually.
+     */
+    this.abortSignal.abort();
+  }
+
+  async done(): Promise<ServiceOutputTypes> {
+    return await Promise.race([this.__doMultipartUpload(), this.__abortTimeout(this.abortSignal.signal)]);
+  }
+
+  on(event: "httpUploadProgress", listener: (progress: Progress) => void): any {
+    this.uploadEvent = event;
+    super.on(event, listener);
+  }
+
+  async __doConcurrentUpload(dataFeeder: AsyncGenerator<RawDataPart>, id: number): Promise<void> {
+    for await (const dataPart of dataFeeder) {
+      if (this.uploadedParts.length > this.MAX_PARTS) {
+        throw new Error(
+          `Exceeded ${this.MAX_PARTS} as part of the upload to ${this.params.Key} and ${this.params.Bucket}.`
+        );
+      }
+
+      try {
+        let partResult = await this.client.send(
+          new UploadPartCommand({
+            ...this.params,
+            UploadId: this.uploadId,
+            Body: dataPart.data,
+            PartNumber: dataPart.partNumber,
+          })
+        );
+        if (this.abortSignal.signal.aborted) {
+          return;
```

Comment:
This line should throw AbortError
```suggestion
          throw Object.assign(new Error("Upload aborted."), { name: "AbortError" });
```

Reply:
I dont think each of the concurrent uploaders should throw an abort error. This return is here as part of clean up. Otherwise we would propagate 5 errors to the user by default.

4 from the 4 default concurrent uploaders and 1 from the main multipart upload manager.

Reply:
actually.... doing some testing .... If we reject, the abort Timeout throws an error to the parent of Upload. I dont think we should throw again here. 
1. This branch is a shortcut to clean up once the the uploader has entered an abort state
2. Its not possible for the top level `Upload.done()` to throw multiple times (i'm 99% sure - and if it threw more than once for a single abort that would be scary too).  Since we throw from abortTimeout and thats triggered immediately in the top level promise race the exception will always come from that abortTimeout function. Errors thrown additional times from inside the concurrent uploader or the main uploader wont ever been seen. They may as well just return in my opinion. No reason I see to throw the abort error from multiple places. Especially if those secondary places have no way of making their error visible.

What do you think? @trivikr

Reply:
Spoke offline with trivikr about this. We'll only be throwing from the abortTimeout and using checks here to clean up promises in flight.

Reply:
Addressed